homeassistant/components/sensor/yr.py

```diff
@@ -153,50 +151,60 @@ def __init__(self, hass, coordinates, forecast, devices):
 
+        if self._unsubscribe_download_new_data:
+            self._unsubscribe_download_new_data()
+            self._unsubscribe_download_new_data = None
+
         def try_again(err: str):
-            """Retry in 15 minutes."""
-            _LOGGER.warning("Retrying in 15 minutes: %s", err)
-            self._nextrun = None
-            nxt = dt_util.utcnow() + timedelta(minutes=15)
-            if nxt.minute >= 15:
-                async_track_point_in_utc_time(self.hass, self.async_update,
-                                              nxt)
-
-        if self._nextrun is None or dt_util.utcnow() >= self._nextrun:
-            try:
-                websession = async_get_clientsession(self.hass)
-                with async_timeout.timeout(10, loop=self.hass.loop):
-                    resp = yield from websession.get(
-                        self._url, params=self._urlparams)
-                if resp.status != 200:
-                    try_again('{} returned {}'.format(resp.url, resp.status))
-                    return
-                text = yield from resp.text()
-
-            except (asyncio.TimeoutError, aiohttp.ClientError) as err:
-                try_again(err)
+            """Retry in 15 to 20 minutes."""
+            minutes = 15 + randrange(6)
+            _LOGGER.error("Retrying in %i minutes: %s", minutes, err)
+            async_call_later(self.hass, minutes*60, self.download_new_data)
+        try:
+            websession = async_get_clientsession(self.hass)
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                resp = yield from websession.get(
+                    self._url, params=self._urlparams)
+            if resp.status != 200:
+                try_again('{} returned {}'.format(resp.url, resp.status))
                 return
+            text = yield from resp.text()
 
-            try:
-                self.data = xmltodict.parse(text)['weatherdata']
-                model = self.data['meta']['model']
-                if '@nextrun' not in model:
-                    model = model[0]
-                self._nextrun = dt_util.parse_datetime(model['@nextrun'])
-            except (ExpatError, IndexError) as err:
-                try_again(err)
-                return
+        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
+            try_again(err)
+            return
+
+        try:
+            self.data = xmltodict.parse(text)['weatherdata']
+        except (ExpatError, IndexError) as err:
+            try_again(err)
+            return
+
+        _unsub = async_track_utc_time_change(self.hass,
+                                             self.download_new_data,
+                                             minute=self._update_time[0],
+                                             second=self._update_time[1])
+        self._unsubscribe_download_new_data = _unsub
+
+        yield from self.async_update()
+
+    @asyncio.coroutine
+    def async_update(self, *_):
```

Comment:
This should not be called async_update. `async_update` is the method that Home Assistant would call if it was a polled device. This makes it confusing (and might break in the future because Home Assistant sees a method that means something else)